ratis-server/src/main/java/org/apache/ratis/server/storage/SnapshotManager.java

```diff
@@ -55,10 +58,22 @@ public class SnapshotManager {
   private static final String TMP = ".tmp";
 
   private final RaftPeerId selfId;
+
+  private final Supplier<File> snapshotDir;
+  private final Supplier<File> tmp;
+  private final Function<FileChunkProto, String> getRelativePath;
   private final Supplier<MessageDigest> digester = JavaUtils.memoize(MD5Hash::getDigester);
 
-  SnapshotManager(RaftPeerId selfId) {
+  SnapshotManager(RaftPeerId selfId, Supplier<RaftStorageDirectory> dir, StateMachineStorage smStorage) {
     this.selfId = selfId;
+    this.snapshotDir = MemoizedSupplier.valueOf(
+        () -> Optional.ofNullable(smStorage.getSnapshotDir()).orElseGet(() -> dir.get().getStateMachineDir()));
+    this.tmp = MemoizedSupplier.valueOf(
+        () -> Optional.ofNullable(smStorage.getTmpDir()).orElseGet(() -> dir.get().getTmpDir()));
+
+    this.getRelativePath =
+        c -> dir.get().getStateMachineDir().toPath()
+            .relativize(new File(dir.get().getRoot(), c.getFilename()).toPath()).toString();
```

Comment:
Use a MemoizedSupplier so that the sm dir path don't have to be computed in every call:
```java
    final Supplier<Path> smDir = MemoizedSupplier.valueOf(() -> dir.get().getStateMachineDir().toPath());
    this.getRelativePath = c -> smDir.get().relativize(
        new File(dir.get().getRoot(), c.getFilename()).toPath()).toString();
```

Reply:
Got it. Done.